File: RangeSumQuery-Mutable4.cpp
#include<iostream>
#include<vector>
#include<math.h>

using namespace std;

class SegmentTreeNode {
public:
	int start, end;
	SegmentTreeNode *left;
	SegmentTreeNode *right;
	int sum;
	
	SegmentTreeNode(int start, int end) : start(start), end(end), left(NULL), right(NULL) {
		this->start = start;
		this->end = end;
		this->sum = 0;
	}
};

class NumArray {
private:
	SegmentTreeNode *root;
public:
    NumArray(vector<int> nums) {
    	if(nums.size() > 0) {
    		root = buildTree(nums, 0, nums.size() - 1);
		}
    }
    
    SegmentTreeNode* buildTree(vector<int> nums, int start, int end) {
    	if(start > end) {
    		return NULL; 
		}else {
			SegmentTreeNode *node = new SegmentTreeNode(start, end);
			if(start == end) {
				node->sum = nums[start];
			}else {
				int mid = start + (end - start) / 2;
				node->left = buildTree(nums, start, mid);
				node->right = buildTree(nums, mid + 1, end);
				node->sum = node->left->sum + node->right->sum;
			}
			return node;
		}
	}
    
    void update(int i, int val) {
        update(root, i, val);
    }
    
    void update(SegmentTreeNode *root, int i, int val) {
    	if(root->start == root->end) {
    		root->sum = val;
		}else {
			int mid = root->start + (root->end - root->start) / 2;
			if(i <= mid) {
				update(root->left, i, val);
			}else {
				update(root->right, i, val);
			}
			
			root->sum = root->left->sum + root->right->sum;
		}
	}
    
    int sumRange(int i, int j) {
        return sumRange(root, i, j);
    }
    
    int sumRange(SegmentTreeNode *root, int i, int j) {
    	if(root->start == i && root->end == j) {
    		return root->sum;
		}
		int mid = root->start + (root->end - root->start) / 2;
		if(i >= mid + 1) return sumRange(root->right, i, j);
		else if(j <= mid) return sumRange(root->left, i, j);
		else return sumRange(root->left, i, mid) + sumRange(root->right, mid + 1, j);
	}
};

int main() {
	
	int n;
	cin>>n;
	
	vector<int> nums;
	for(int i = 0;i < n;i++) {
		int num;
		cin>>num;
		nums.push_back(num);
	}
	
	NumArray *arr = new NumArray(nums);
	cout<<arr->sumRange(0, 2)<<endl;
	arr->update(1, 2);
	cout<<arr->sumRange(0, 2)<<endl;
	
	return 0;
	
}

File: RangeSumQuery-Mutable4.py
class SegmentTreeNode:
    def __init__(self, i, j):
        self.i = i
        self.j = j
        self.left = None
        self.right = None
        self.sum = 0

class NumArray:

    def __init__(self, nums):
        """
        :type nums: List[int]
        """
        if not nums:
            return
        self.root = self.build(nums, 0, len(nums) - 1)

    def build(self, nums, begin, end):
        if begin > end:
            return None
        root = SegmentTreeNode(begin, end)
        if begin == end:
            root.sum = nums[begin]
        else:
            mid = begin + (end - begin) // 2
            root.left = self.build(nums, begin, mid)
            root.right = self.build(nums, mid + 1, end)
            root.sum = root.left.sum + root.right.sum
        return root

    def update(self, i, val):
        """
        :type i: int
        :type val: int
        :rtype: void
        """
        def update(root, i, val):
            if root.i == root.j:
                root.sum = val
            else:
                mid = root.i + (root.j - root.i) // 2
                if i > mid:
                    update(root.right, i, val)
                else:
                    update(root.left, i, val)
                root.sum = root.left.sum + root.right.sum
        update(self.root, i, val)

    def sumRange(self, i, j):
        """
        :type i: int
        :type j: int
        :rtype: int
        """
        def sumRange(root, i, j):
            if root.i == i and root.j == j:
                return root.sum
            mid = root.i + (root.j - root.i) // 2
            if i > mid:
                return sumRange(root.right, i, j)
            elif j <= mid:
                return sumRange(root.left, i, j)
            else:
                return sumRange(root.left, i, mid) + sumRange(root.right, mid + 1, j)
        return sumRange(self.root, i, j)


if __name__ == '__main__':
    nums = [int(num) for num in input().split()]
    arr = NumArray(nums)
    print(arr.sumRange(0, 2))
    arr.update(1, 2)
    print(arr.sumRange(0, 2))
